std-bits/table/src/main/java/org/enso/table/parsing/NumberParser.java

```diff
@@ -196,259 +26,132 @@ private static Pattern buildPattern(
-   */
-  private Pattern patternForIndex(int index) {
-    int allowedSet = (allowCurrency ? ALLOWED_CCY_PATTERNS : ALLOWED_NON_CCY_PATTERNS);
-    int separatorsIndex = index / allowedSet;
-    int patternIndex = index % allowedSet;
-
-    if (separatorsIndex >= SEPARATORS.length) {
-      return null;
-    }
-
-    return buildPattern(
-        allowDecimal,
-        allowCurrency,
-        allowScientific,
-        trimValues,
-        patternIndex,
-        SEPARATORS[separatorsIndex]);
+    var numberWithSeparators = NumberWithSeparators.fromSeparators(thousandSeparator, decimalPoint);
+    this.parser =
+        new FormatDetectingNumberParser(
+            allowSymbol,
+            allowLeadingZeroes,
+            allowLeadingTrailingWhitespace,
+            NegativeSign.UNKNOWN,
+            numberWithSeparators);
   }
 
   @Override
-  public Object parseSingleValue(String text, ParseProblemAggregator problemAggregator) {
-    int index = 0;
-    var pattern = patternForIndex(index);
-    while (pattern != null) {
-      var value = innerParseSingleValue(text, pattern);
-      if (value != null) {
-        return value;
-      }
-
-      index++;
-      pattern = patternForIndex(index);
-    }
-
-    problemAggregator.reportInvalidFormat(text);
-    return null;
+  protected Builder makeBuilderWithCapacity(int capacity, ProblemAggregator problemAggregator) {
+    return isInteger
+        ? NumericBuilder.createLongBuilder(capacity, integerTargetType, problemAggregator)
+        : NumericBuilder.createDoubleBuilder(capacity, problemAggregator);
   }
 
   @Override
   public Storage<?> parseColumn(
       Storage<String> sourceStorage, CommonParseProblemAggregator problemAggregator) {
-    int index = 0;
-    var pattern = patternForIndex(index);
-
-    int bestIndex = 0;
-    int bestCount = -1;
-    while (pattern != null) {
-      ProblemAggregator inner = problemAggregator.createSimpleChild();
-      Builder builder = makeBuilderWithCapacity(sourceStorage.size(), inner);
-      int failedAt = parseColumnWithPattern(pattern, sourceStorage, builder, null);
-      if (failedAt == -1) {
-        return builder.seal();
-      }
+    Builder builder =
+        makeBuilderWithCapacity(sourceStorage.size(), problemAggregator.createSimpleChild());
 
-      // If there was a failure, we abandon this branch - thus we discard any problems that might
-      // have been reported by the inner aggregator.
-      inner.detachFromParent();
+    var context = Context.getCurrent();
+    for (int i = 0; i < sourceStorage.size(); i++) {
+      var text = sourceStorage.getItemBoxed(i);
 
-      if (failedAt > bestCount) {
-        bestCount = failedAt;
-        bestIndex = index;
+      // Check if in unknown state
+      var isInMixedState =
+          !isInteger
+              && (parser.numberWithSeparators() == NumberWithSeparators.DOT_UNKNOWN
+                  || parser.numberWithSeparators() == NumberWithSeparators.COMMA_UNKNOWN);
```

Comment:
This check is IMO crossing the layers of abstraction and leaking the parser internals too much.

If I understand correctly, now `NumberParser` is supposed to be a kind of adapter that wraps the `FormatDetectingNumberParser` to be usable by Table's common `IncrementalDatatypeParser` API.

While not a big issue, it feels like this should be abstracted into higher level methods on the `FormatDetectingNumberParser`.

e.g.
```suggestion
      var isInMixedState =
          !isInteger
              && parsed.hasEncounteredAnyDecimalSeparator();
```

Reply:
This way the lower level handling of separator detection is hidden and the public API of `FormatDetectingNumberParser` only exposes the high-level concepts of parser state instead of internals.

I'd go even further and make `numberWithSeparators` private and expose 2 getters on the parser: `Char detectedThousandsSeparator()` and `detectedDecimalSeparator` - that will return the separator if it is already known, or `null` for `UNKNOWN`.

Reply:
What does `isInMixedState` mean?

Reply:
This is a deliberate exposure to allow the detection of when we need to rescan. Only happens when we had a European style kid identified as a US one.

Reply:
This is a deliberate exposure to allow the detection of when we need to rescan. Only happens when we had a European style kid identified as a US one.

Reply:
NumbersWithSeparstors was the output enumerator type from the parser. It exposes the specific separators but as we only support a specific list seemed better than pretending we support any.

Reply:
> Only happens when we had a European style kid identified as a US one.

Would be worth adding a comment explaining exactly that. Otherwise it is not really clear what this state means. It is clear once you know the intricacies of the parser. But if any of us (or a new person) comes here after a year to fix some issue with the parser, I don't think it will be clear.

Reply:
I'm not saying we should not check the parser state here. Just saying that it may be clearer if instead of checking internals for a set of enum values, we could query a getter function that will do the check.

1. This way the function can be given a name - making it easier to understand what is being checked here.
2. The internals get more encapsulated. If the parser gets adds another supported variant of the separators, we may forget to update this code, whereas if the getter lives in the parser class it will be much easier to remember to update it accordingly.

Reply:
> that will return the separator if it is already known, or `null` for `UNKNOWN`.

Looking back at it, this suggestion is wrong. If anything, it should return a `char` and just return `Constants.UNKNOWN` etc. to be able to distinguish it from `Constants.NONE`.